### PR TITLE
[REVIEW] Add pandas nullable type support in `Index.to_pandas`

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description

Resolves: #12216

This PR adds support for `nullable` parameter to `Index.to_pandas`. When `nullable=True`, the corresponding dtype of the returned index object will try to have a pandas nullable dtype if one exists.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/_base_index.py [hunk 1]
```diff
@@ -626,10 +626,18 @@ def notna(self):
...
         raise NotImplementedError
 
-    def to_pandas(self):
+    def to_pandas(self, nullable=False):
         """
         Convert to a Pandas Index.
 
+        Parameters
+        ----------
+        nullable : Boolean, Default False
```
Comment (shwina) [inline python/cudf/cudf/core/_base_index.py:635]:
```suggestion
        nullable : bool, Default False
```

File: python/cudf/cudf/core/_base_index.py [hunk 2]
```diff
@@ -626,10 +626,18 @@ def notna(self):
...
+            If ``nullable`` is ``True``, the resulting series will be
+            having a corresponding nullable Pandas dtype. If ``nullable``
+            is ``False``, the resulting series will either convert null
+            values to ``np.nan`` or ``None`` depending on the dtype.
```
Comment (shwina) [inline python/cudf/cudf/core/_base_index.py:639]:
```suggestion
            If ``nullable`` is ``True``, the resulting index will be
            having a corresponding nullable Pandas dtype. If ``nullable``
            is ``False``, the resulting index will either convert null
            values to ``np.nan`` or ``None`` depending on the dtype.
```

Comment (galipremsagar) [reply python/cudf/cudf/core/_base_index.py]: Yikes, apologies for overlooking the text here.

**Review by shwina — APPROVED** [review]

File: python/cudf/cudf/core/_base_index.py [hunk 3]
```diff
@@ -626,10 +626,18 @@ def notna(self):
...
-    def to_pandas(self):
+    def to_pandas(self, nullable=False):
         """
         Convert to a Pandas Index.
 
+        Parameters
+        ----------
+        nullable : bool, Default False
+            If ``nullable`` is ``True``, the resulting index will be
+            having a corresponding nullable Pandas dtype. If ``nullable``
```
Comment (vyasr) [inline python/cudf/cudf/core/_base_index.py:637]:
```suggestion
            If ``nullable`` is ``True``, the resulting index will have
            a corresponding nullable Pandas dtype. If ``nullable``
```

**Review by vyasr — COMMENTED** [review]
Rather than silently producing non-nullable dtypes in cases where no pandas nullable equivalent exists, should we raise an error? If pandas ever introduces additional nullable dtypes the current approach would lead to silently creating indexes of the wrong dtype. A loud failure would probably be easier to debug.

**Comment by galipremsagar** [comment]
> Rather than silently producing non-nullable dtypes in cases where no pandas nullable equivalent exists, should we raise an error if nullable is True? If pandas ever introduces additional nullable dtypes the current approach would lead to silently creating indexes of the wrong dtype. A loud failure would probably be easier to debug.

The tricky part is in case of a dataframe with a mix of dtypes that would end up being pandas nullable & non-nullable dtypes: https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/core/dataframe.py#L5037 i.e., we'd be making the following throw an error:

```python

In [2]: s = cudf.Series([1, 2, 3, 4], dtype='timedelta64[ns]')


In [7]: df = cudf.DataFrame({
   ...:    'time': s,
   ...:     'style': ['cup', 'cup', 'cup', 'pack'],
   ...: })

In [8]: df
Out[8]: 
                       time style
0 0 days 00:00:00.000000001   cup
1 0 days 00:00:00.000000002   cup
2 0 days 00:00:00.000000003   cup
3 0 days 00:00:00.000000004  pack

In [9]: df.to_pandas(nullable=True)
Out[9]: 
                       time style
0 0 days 00:00:00.000000001   cup
1 0 days 00:00:00.000000002   cup
2 0 days 00:00:00.000000003   cup
3 0 days 00:00:00.000000004  pack
```

**Comment by galipremsagar** [comment]
> I lean towards the latter since it seems pretty reasonable for users to use this op to convert a DataFrame where only a subset of columns convert cleanly

Yeah, I lean toward 2 too. The reason being with a warning, it might end up being too noisy.

**Comment by galipremsagar** [comment]
> For all other Index types it's well-defined whether such conversion is valid.

By types, I mean't dtypes..`datetime` & `timedelta` dtypes still don't have corresponding nullable-dtypes. So `Series`/`Index`/`DataFrame`/`MultiIndex` will all have similar behavior when one of these dtypes are present.

**Comment by vyasr** [comment]
Right, I just mean that except for DataFrame/MultiIndex there's no mixing. If a Series/Index has types that don't have a nullable pandas dtype, there's no partial conversion to perform. In any case, I'm still fine with the behavior, just wanted to clarify so that the docstring is accurate.

**Review by vyasr — APPROVED** [review]

**Comment by galipremsagar** [comment]
> Right, I just mean that except for DataFrame/MultiIndex there's no mixing. If a Series/Index has types that don't have a nullable pandas dtype, there's no partial conversion to perform. In any case, I'm still fine with the behavior, just wanted to clarify so that the docstring is accurate.

Updated docstring, @vyasr could you take a look at the changes before we merge?

**Comment by vyasr** [comment]
Yup looks good to me, thanks Prem!

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**